src/model/UserphraseModel.h

```diff
@@ -59,20 +59,23 @@ class UserphraseModel final: public QAbstractListModel {
     void addNewPhraseCompleted(const Userphrase& userphrase);
     void removePhraseCompleted(size_t count);
     void refreshCompleted(size_t count);
-    void addNewPhraseFailed();
+    void addNewPhraseFailed();    
+    void undoCompleted(size_t count);
 
 public slots:
     void refresh();
     void add(std::shared_ptr<QString> phrase, std::shared_ptr<QString> bopomofo);
     void importUserphrase(std::shared_ptr<UserphraseImporter> importer);
     void exportUserphrase(std::shared_ptr<UserphraseExporter> exporter);
     void remove(QModelIndexList indexList);
+    void undo();
 
 private:
     void add(const QString &phrase, const QString &bopomofo);
     QString checkBopomofo(const QString &bopomofo) const;
 
     std::unique_ptr<ChewingContext, void (*)(ChewingContext*)> ctx_;
     UserphraseSet userphrase_;
+    UserphraseSet userphrase_tmp_;
```

Comment:
@wcchang1115 : Even "shadow" or "placeholder" is better than unary `tmp`. Every member data in class UserphraseModel should be referred effectively.

By the way, since I am the collaborator, call me @jserv.